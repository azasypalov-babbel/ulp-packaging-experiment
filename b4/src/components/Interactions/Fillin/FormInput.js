import React, { useState, useEffect, useRef, useImperativeHandle, useContext, useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { log } from '../../../lib/logging';
import { isInteractionKey } from '../../../lib/keyboardEvents';

import useGapWidth from '../shared/Write/useGapWidth';
import { insertText } from '../shared/Write/helpers';

import performAfterScroll from '../../../lib/performAfterScroll';
import { ServiceContext } from '../../shared/withServices';
import { useFocusTracking } from '../../shared/hooks/useFocusTracking';
import { isiOS } from '../../../lib/features';
import { useSelector } from 'react-redux';

export const KEYCODES = {
  ENTER: 'Enter',
  ESC: 'Escape'
};

const shouldAutoCapitalize = (targetText = '') =>
  targetText.toUpperCase()[0] === targetText[0];

const ensureFocus = (el) => {
  if (document.activeElement !== el) {
    log('FormInput: Focussing');
    el.focus({ preventScroll: true });
  } else {
    log('FormInput: Already has focus');
  }
};

const sanitizedValue = (text) => {
  // removes leading white-space
  // trailing white-space is kept (it may be a correct word boundary)
  let sanitized = text.replace(/^\s+/g, '');

  // replaces different white-spaces with a white-space ' '
  // removes duplicates
  sanitized = sanitized.replace(/\s+/g, ' ');

  return sanitized;
};

const StyledInput = styled.input`
  width: ${({ gapWidth }) => gapWidth};
`;

const FormInput = React.forwardRef(
  (
    {
      attempt,
      controlled,
      onReturn,
      onLatestTextChange = () => {},
      onKeyPress = () => {},
      targetText,
      alternativeText,
      onBlur = () => {},
      ...props
    },
    ref
  ) => {
    const { number, pending } = attempt;

    const inputRef = useRef();
    const { onBlur: trackFocus } = useFocusTracking(inputRef);
    const isZendeskWidgetOpen = useSelector(state => state.zendeskWidget.isOpen);

    const refocus = () => {
      setTimeout(() => {
        inputRef.current?.focus();
      }, 1)
    }

    const handleBlur = useCallback((...args) => {
      onBlur(...args);
      trackFocus(...args);
      if (!isBlurredWithEnter && !isZendeskWidgetOpen) {
        refocus();
      } else {
        setIsBlurredWithEnter(false);
      }
    }, [onBlur, trackFocus, isBlurredWithEnter, isZendeskWidgetOpen]);

    useEffect(() => {
      if (!isZendeskWidgetOpen) {
        refocus();
      }
    }, [isZendeskWidgetOpen])

    const [editable, setEditable] = useState(true);
    const [isBlurredWithEnter, setIsBlurredWithEnter] = useState(false);
    const isFocused = () => document.activeElement === inputRef?.current;

    const handleChange = (event) => {
      onLatestTextChange(event.target.value);
    };

    const triggerReturn = () => {
      const text = sanitizedValue(inputRef.current.value);
      onReturn(text);
    };

    const triggerSolve = () => {
      onReturn(targetText);
    };

    const handleKeyDown = (event) => {
      if (isInteractionKey(event, ['Shift+Escape'])) {
        event.preventDefault();
        triggerSolve();
      }
    };

    const handleKeyPress = (event) => {
      if (event.key === KEYCODES.ENTER && !event.repeat) {
        event.preventDefault();
        setIsBlurredWithEnter(true);
        setTimeout(() => {
          inputRef.current?.blur();
        }, 1);
        triggerReturn();
        return;
      }
      onKeyPress(event);
    };

    useImperativeHandle(ref, () => ({
      focus: () => {
        ensureFocus(inputRef.current);
      },
      retry: () => {
        log('FormInput: retry');
        ensureFocus(inputRef.current);
      },
      insert: (text) => {
        log('FormInput: insert');
        if (!editable) {
          console.warn('Gap is not editable');
          return;
        }

        ensureFocus(inputRef.current);
        insertText(text);
      },
      delete: () => {
        log('FormInput: delete');
        if (!editable) {
          console.warn('Gap is not editable');
          return;
        }

        ensureFocus(inputRef.current);

        const caretIndex = inputRef.current.selectionStart;
        const hasSelection = caretIndex !== inputRef.current.selectionEnd;

        // Disable selection so it's not visible to the user while deletion is made

        inputRef.current.classList.add('no-select');

        if (hasSelection) {
          inputRef.current.setSelectionRange(
            caretIndex,
            inputRef.current.selectionEnd
          );
        } else {
          inputRef.current.setSelectionRange(caretIndex - 1, caretIndex);
        }

        insertText('');
        inputRef.current.classList.remove('no-select');
      },
      deleteAll: () => {
        log('FormInput: deleteAll');

        if (!editable) {
          console.warn('Gap is not editable');
          return;
        }

        ensureFocus(inputRef.current);

        inputRef.current.setSelectionRange(0, inputRef.current.value.length);

        insertText('');
      },
      return: () => {
        log('FormInput: return');
        if (!editable) {
          console.warn('Gap is not editable');
          return;
        }
        triggerReturn();
      },
      addEventListener: (...args) => {
        return inputRef?.current?.addEventListener.apply(null, args);
      },
      removeEventListener: (...args) => {
        return inputRef?.current?.removeEventListener.apply(null, args);
      },
      isFocused,
      getValue: () => inputRef?.current?.value
    }));

    const { focusService } = useContext(ServiceContext);
    // Initial focus
    useEffect(() => {
      inputRef.current.blur();

      const cleanup = performAfterScroll(() => {
        inputRef.current.focus({ preventScroll: true });
      });

      return () => {
        cleanup();
      };
    }, [focusService]);

    // iOS caret increases the required gap width
    const adjustForCaret = isiOS() && (!number || isFocused());

    // Set width
    const gapWidth = useGapWidth({ ref: inputRef, targetText, adjustForCaret });

    // Block editing when pending
    useEffect(() => {
      setEditable(!pending);
    }, [pending]);

    return (
      // Having form[action] gives the blue Return button on iPad (however blue lablled "Open" button on iPhone)
      <form action={''} onSubmit={(e) => e.preventDefault()}>
        <StyledInput
          {...props}
          onBlur={handleBlur}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
          data-focus-on-key="Shift+Escape"
          onKeyPress={handleKeyPress}
          autoCapitalize={shouldAutoCapitalize(targetText) ? 'on' : 'off'}
          autoComplete={'off'}
          autoCorrect={'off'}
          spellCheck={'false'}
          readOnly={controlled}
          ref={inputRef}
          role={'textbox'}
          data-solution={targetText}
          data-solution-alternative={alternativeText}
          data-textbox-type={'input'}
          data-selector={'active-gap'}
          gapWidth={gapWidth}
        />
      </form>
    );
  }
);

FormInput.displayName = 'FormInput';

FormInput.propTypes = {
  attempt: PropTypes.object.isRequired,
  controlled: PropTypes.bool.isRequired,
  onReturn: PropTypes.func.isRequired,
  onLatestTextChange: PropTypes.func,
  onBlur: PropTypes.func,
  onKeyPress: PropTypes.func,
  targetText: PropTypes.string.isRequired,
  alternativeText: PropTypes.string
};

export default FormInput;
