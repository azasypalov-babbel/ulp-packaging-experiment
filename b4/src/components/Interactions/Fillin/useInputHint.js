import { useState, useCallback } from 'react';

export const HINT_LEVEL = {
  NONE: 'NONE',
  TRANSLATION: 'TRANSLATION',
  HINT: 'HINT',
  SOLUTION: 'SOLUTION'
};

const hintSequence = [
  HINT_LEVEL.NONE,
  // HINT_LEVEL.TRANSLATION,
  HINT_LEVEL.HINT,
  HINT_LEVEL.SOLUTION
];

const getNextHintLevel = (hintLevel) =>
  hintSequence[hintSequence.indexOf(hintLevel) + 1] || HINT_LEVEL.NONE;

export const useInputHint = () => {
  const [hintLevel, setHintLevel] = useState(HINT_LEVEL.NONE);
  const [nextHintLevel, setNextHintLevel] = useState(
    getNextHintLevel(HINT_LEVEL.NONE)
  );

  const handleHintClick = useCallback(() => {
    setNextHintLevel((nextHintLevel) => {
      setHintLevel(nextHintLevel);
      return getNextHintLevel(nextHintLevel);
    });
  }, []);

  const resetHint = useCallback(() => {
    setHintLevel(HINT_LEVEL.NONE);
    setNextHintLevel(getNextHintLevel(HINT_LEVEL.NONE));
  }, []);

  return {
    resetHint,
    hintLevel,
    nextHintLevel,
    onHint: handleHintClick
  };
};
