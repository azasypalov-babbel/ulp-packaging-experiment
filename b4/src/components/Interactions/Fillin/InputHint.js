import React, { useState } from 'react';
import styled, { keyframes } from 'styled-components';
import posed, { PoseGroup } from 'react-pose';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { track } from '../../../dux/tracker/actions';
import shuffle from '../../../lib/shuffle';
import Text from '../../shared/Text';
import { splitBefore, trimWordStart } from '../shared/Write/helpers';
import { useMapUsedTokens } from '../shared/Write/useMapUsedTokens';

import { HINT_LEVEL } from './useInputHint';

const enterKeyframes = keyframes`
  from {
    opacity: 0;
    transform: translateX(-50%) translateY(10%);
  }
  to {
    opacity: 1;
    transform: translateX(-50%) translateY(0);
  }
`;

const StyledCharacter = styled(posed.div({}))`
  opacity: ${(props) => props.used ? 0.5 : 1};
  transition: order 0.25s;
`;

const StyledInputHint = styled.div`
  position: absolute;
  user-select: none;

  display: flex;

  bottom: 100%;
  left: 50%;
  transform: translateX(-50%);

  animation: ${enterKeyframes} 0.2s;
`;

const StyledHintWord = styled.div`
  position: relative;

  font-size: 1rem;
  white-space: nowrap;

  display: flex;

  padding: 0.5em;
  margin: 0.5em 0;

  background-color: ${((props) => props.theme.cascada.white)};
  border-radius: 0.25em;

  border: 1px solid ${(props) => props.theme.cascada.pastelStorm};
  box-shadow: 0 0 0.2rem ${(props) => props.theme.cascada.silver};

  & + & {
    margin-left: 0.5rem;
  }
`;

const trackHintClick = (track) => {
  /* eslint-disable camelcase */
  track({
    event: 'gui:interacted',
    version: 1,
    payload: {
      gui_container: 'input_hint',
      gui_element: 'hint_tooltip',
      interaction: 'clicked',
      origin: 'lesson_player'
    }
  });
  /* eslint-enable camelcase */
};

export const InputHint = ({
  latestText,
  targetText,
  hintLevel,
  track,
  learnLanguageAlpha3
}) => {
  const [seed] = useState(Math.random() * 10000);

  const mapUsedTokens = useMapUsedTokens(learnLanguageAlpha3);

  if (hintLevel === HINT_LEVEL.NONE) {
    return null;
  }

  const usedWordMap = splitBefore(
    mapUsedTokens({
      inputText: latestText,
      referenceTokens: targetText.split('')
    }),
    ({ token }) => token === ' '
  );

  const hintWords = hintLevel === HINT_LEVEL.HINT
    ? usedWordMap.map((word) => shuffle({ array: trimWordStart(word), seed, notOriginalOrder: true }))
    : usedWordMap;

  return (
    <StyledInputHint
      data-selector="input-hint"
      onClick={() => trackHintClick(track)}
    >
      {hintWords.map((hintCharacters, index) =>
        <StyledHintWord key={index}>
          <PoseGroup>
            {hintCharacters.map(({ used, token, id }) =>
              <StyledCharacter
                data-selector="input-hint-char"
                key={id}
                used={used}>
                <Text>{token}</Text>
              </StyledCharacter>
            )}
          </PoseGroup>
        </StyledHintWord>
      )}
    </StyledInputHint>
  );
};

const mapDispatchToProps = {
  track
};

InputHint.propTypes = {
  learnLanguageAlpha3: PropTypes.string.isRequired,
  latestText: PropTypes.string.isRequired,
  targetText: PropTypes.string.isRequired,
  hintLevel: PropTypes.string.isRequired,
  track: PropTypes.func.isRequired
};

export default connect(null, mapDispatchToProps)(InputHint);
