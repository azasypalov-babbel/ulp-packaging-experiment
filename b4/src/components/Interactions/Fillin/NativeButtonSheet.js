import React from 'react';
import PropTypes from 'prop-types';
import { useEffect, useContext } from 'react';
import { connect } from 'react-redux';

import { track } from '../../../dux/tracker/actions';
import { ServiceContext } from '../../shared/withServices';
import { BRIDGE_EVENTS } from '../../../services/nativeHintButton';
import { HINT_LEVEL } from './useInputHint';
import { renderInBottomLayout } from '../../shared/BottomLayout';
import { compose } from 'redux';

const NativeHintButton = ({ onClick, nextHintLevel }) => {
  const { nativeHintButton } = useContext(ServiceContext);

  useEffect(() => {
    nativeHintButton.addEventListener(BRIDGE_EVENTS.ON_CLICK, onClick);
    return () => {
      nativeHintButton.removeEventListener(BRIDGE_EVENTS.ON_CLICK, onClick);
    };
  }, [nativeHintButton, onClick]);

  useEffect(() => {
    nativeHintButton.show(nextHintLevel);
    return () => {
      nativeHintButton.hide();
    };
  }, [nextHintLevel, nativeHintButton]);

  // Reserve some height in the DOM to represent the button
  // The actual UI is rendered natively so this height may
  // Need to be adjusted if the native UI changes
  return <div style={{ height: '5em' }}/>;
};

NativeHintButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  nextHintLevel: PropTypes.oneOf(Object.values(HINT_LEVEL))
};


const trackGuiInteractedEvent = (track, guiElement) => {
  /* eslint-disable camelcase */
  track({
    event: 'gui:interacted',
    version: 1,
    payload: {
      gui_container: 'native_hint_button',
      gui_element: guiElement,
      interaction: 'clicked',
      origin: 'lesson_player'
    }
  });
  /* eslint-enable camelcase */
};

const trackButtonClicked = (track, hintLevel) => {
  if (hintLevel === 'HINT') {
    trackGuiInteractedEvent(track, 'see_hint_button');
  } else if (hintLevel === 'SOLUTION') {
    trackGuiInteractedEvent(track, 'see_solution_button');
  }
};

const NativeButtonSheet = ({ onHint, nextHintLevel, track }) => {
  if (nextHintLevel === HINT_LEVEL.NONE) return null;

  return (<NativeHintButton
    onClick={() => {
      trackButtonClicked(track, nextHintLevel);
      onHint();
    }}
    nextHintLevel={nextHintLevel}
  />);
};

const mapDispatchToProps = {
  track
};

NativeButtonSheet.propTypes = {
  onHint: PropTypes.func.isRequired,
  nextHintLevel: PropTypes.string.isRequired,
  track: PropTypes.func.isRequired
};

export default compose(
  connect(null, mapDispatchToProps),
  renderInBottomLayout
)(NativeButtonSheet);
