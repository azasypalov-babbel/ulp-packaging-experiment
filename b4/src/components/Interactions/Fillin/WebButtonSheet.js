import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { track } from '../../../dux/tracker/actions';
import IconButton from '../../shared/IconButton/IconButton';
import withTranslations from '../../shared/withTranslations';
import { renderInBottomLayout } from '../../shared/BottomLayout';
import { HINT_LEVEL } from './useInputHint';

const StyledWebButtonSheet = styled.div`
  display: flex;
  flex-direction: row;
  padding: 0.75rem;
  justify-content: space-between;

  order: 100;

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    padding: 1.5rem;
  }
`;

const StyledIconButton = styled(IconButton)`
  height: 2.5rem;

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    height: 3rem;
  }
`;

export const WebButtonSheet = ({
  onTransliterationToggled,
  isTransliterationTableSupported,
  nextHintLevel,
  translations,
  onHint,
  track
}) => {
  const handleTransliterationToggleClick = () => {
    /* eslint-disable camelcase */
    track({
      event: 'gui:interacted',
      version: 1,
      payload: {
        gui_element: 'transliterations_toggle',
        interaction: 'clicked',
        origin: 'lesson_player'
      }
    });
    /* eslint-enable camelcase */

    onTransliterationToggled();
  };

  const handleHintClick = () => {
    /* eslint-disable camelcase */
    track({
      event: 'gui:interacted',
      version: 1,
      payload: {
        gui_container: 'web_hint_button',
        gui_element: nextHintLevel === 'HINT' ? 'see_hint_button' : 'see_solution_button',
        interaction: 'clicked',
        origin: 'lesson_player'
      }
    });
    /* eslint-enable camelcase */

    onHint();
  };

  const showTypingTips = isTransliterationTableSupported;
  const showHintButton = nextHintLevel !== HINT_LEVEL.NONE;

  return (
    <StyledWebButtonSheet onClick={(event) => event.stopPropagation()}>
      <div>
        {showTypingTips && <StyledIconButton
          data-selector="transliteration-toggle"
          onClick={handleTransliterationToggleClick}
          iconName="KeyboardIcon">
          {translations.typingTips}
        </StyledIconButton>}
      </div>
      <div>
        {showHintButton && <StyledIconButton
          onClick={handleHintClick}
          iconName={nextHintLevel === 'HINT' ? 'LightBulbIcon' : 'CheckIcon'}
          data-selector={nextHintLevel === 'HINT' ? 'show-hint' : 'show-solution'}>
          {nextHintLevel === 'HINT' ? translations.hint : translations.solution}
        </StyledIconButton>}
      </div>
    </StyledWebButtonSheet>
  );
};

WebButtonSheet.propTypes = {
  onTransliterationToggled: PropTypes.func.isRequired,
  track: PropTypes.func.isRequired,
  isTransliterationTableSupported: PropTypes.bool.isRequired,
  nextHintLevel: PropTypes.string.isRequired,
  onHint: PropTypes.func.isRequired,
  translations: PropTypes.shape({
    typingTips: PropTypes.string.isRequired,
    solution: PropTypes.string.isRequired,
    hint: PropTypes.string.isRequired
  }).isRequired
};

const mapDispatchToProps = {
  track
};

const getTranslations = (translate) => ({
  hint: translate('button.see_hint'),
  solution: translate('button.see_solution'),
  typingTips: translate('button.typing_tips')
});

export default compose(
  connect(null, mapDispatchToProps),
  withTranslations(getTranslations),
  renderInBottomLayout
)(WebButtonSheet);
