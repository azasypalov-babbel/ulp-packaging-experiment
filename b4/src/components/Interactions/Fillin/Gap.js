import React, { useEffect, useMemo, useState, useRef, useImperativeHandle } from 'react';
import PropTypes from 'prop-types';

import * as features from '../../../lib/features';
import useActiveGapStale from '../shared/Write/useActiveGapStale';

import InputHint from './InputHint';

import StyledGap from '../shared/Write/StyledGap';
import FormInput from './FormInput';
import TextGap from '../shared/Write/TextGap';
import { getClosestMatch, getAlternativeSolution } from '../shared/Write/helpers';
import ToolTip from '../../shared/ToolTip';
import { INTERACTION_STATES } from '../shared/scene/useScene';
import { RESULTS } from '../../../lib/matchingUtils/evaluate';

const Gap = React.forwardRef(({
  submitHint,
  interactionState,
  targetChoices,
  attempt,
  onReturn,
  hintLevel,
  handleKeyPress,
  learnLanguageAlpha3,
  toggleTransliterationVisibility,
  shouldShowTransliteration,
  onRetry
}, ref) => {
  const inputRef = useRef();
  const [latestText, setLatestText] = useState('');

  useImperativeHandle(ref, () => ({
    focus: () => {
      inputRef.current?.focus({ preventScroll: true });
    },
    deleteAll: () => {
      inputRef.current?.deleteAll();
      inputRef.current?.retry();
    }
  }));

  const [
    onKeyUp,
    onBlur,
    shouldShowSubmitTip
  ] = useActiveGapStale(!features.isWebview());

  useEffect(() => {
    if (shouldShowTransliteration && shouldShowSubmitTip) {
      toggleTransliterationVisibility();
    }
  }, [
    shouldShowSubmitTip,
    shouldShowTransliteration,
    toggleTransliterationVisibility
  ]);

  const handleReturn = (text) => {
    onReturn(text); 
    onBlur();
  }

  const handleLatestTextChange = (text) => {
    if (interactionState === INTERACTION_STATES.FEEDBACK) {
      onRetry();
    }
    setLatestText(text);
  };

  const targetTexts = useMemo(() => targetChoices
    .filter((choice) => choice.correct)
    .map((choice) => choice.sentence), [targetChoices]);

  const { targetText } = useMemo(() => getClosestMatch({ text: latestText, targetTexts }), [latestText, targetTexts]);
  const alternativeText = getAlternativeSolution(targetTexts, targetText);

  if (interactionState === INTERACTION_STATES.INACTIVE) {
    return (
      <StyledGap active={false}>
        <TextGap
          text={attempt.number ? targetText : ' '}
          targetText={targetText}
          alternativeText={alternativeText}
        />
      </StyledGap>
    );
  }

  const showFeedback = interactionState === INTERACTION_STATES.FEEDBACK;

  if (showFeedback && attempt.solved) {
    const { inputText, text } = attempt;
    const hasTypoFeedback = attempt.feedbackType === RESULTS.TYPO;
    return (
      <StyledGap
        active={true}
        solved={true}
        typo={hasTypoFeedback}>
        <TextGap
          text={hasTypoFeedback ? inputText : text}
          targetText={hasTypoFeedback ? inputText : targetText}
          alternativeText={alternativeText}
        />
      </StyledGap>
    );
  }

  return (
    <StyledGap
      active={true}
      mistaken={!attempt.solved && showFeedback}
    >
      <ToolTip text={submitHint} visible={shouldShowSubmitTip}>
        <InputHint
          latestText={latestText}
          targetText={targetText}
          alternativeText={alternativeText}
          hintLevel={hintLevel}
          learnLanguageAlpha3={learnLanguageAlpha3}
        />
        <FormInput
          ref={inputRef}
          attempt={attempt}
          onKeyUp={onKeyUp}
          onBlur={onBlur}
          controlled={false}
          onLatestTextChange={handleLatestTextChange}
          onKeyPress={handleKeyPress}
          onReturn={handleReturn}
          targetText={targetText}
          alternativeText={alternativeText}
          onClick={onRetry}
        />
      </ToolTip>
    </StyledGap>
  );
});

Gap.displayName = 'Gap';

Gap.propTypes = {
  onRetry: PropTypes.func.isRequired,
  targetChoices: PropTypes.arrayOf(
    PropTypes.shape({
      sentence: PropTypes.string.isRequired,
      correct: PropTypes.bool.isRequired,
      oddSolution: PropTypes.bool.isRequired
    })
  ).isRequired,
  hintLevel: PropTypes.string.isRequired,
  interactionState: PropTypes.string.isRequired,
  attempt: PropTypes.object,
  handleKeyPress: PropTypes.func,
  onReturn: PropTypes.func.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  toggleTransliterationVisibility: PropTypes.func.isRequired,
  shouldShowTransliteration: PropTypes.bool,
  submitHint: PropTypes.string.isRequired
};

export default Gap;
