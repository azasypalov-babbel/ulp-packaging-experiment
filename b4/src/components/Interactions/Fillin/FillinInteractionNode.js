import React, { useRef, useCallback } from 'react';
import PropTypes from 'prop-types';
import * as features from '../../../lib/features';
import { attemptType, interactiveNodeType } from '../shared/scene/scenePropTypes';
import NativeButtonSheet from './NativeButtonSheet';
import WebButtonSheet from './WebButtonSheet';
import { useInputHint } from './useInputHint';
import { isiOS } from '../../../lib/features';
import Text from '../../shared/Text';
import Gap from './Gap';
import FillinFeedback from './FillinFeedback';
import { useTransliterationReplace } from '../shared/Transliteration/replace';
import { useTransliterationUIState } from '../shared/Transliteration/UIState';
import TransliterationTable from '../shared/Transliteration/TransliterationTable';
import { INTERACTION_STATES } from '../shared/scene/useScene';
import withTranslations from '../../shared/withTranslations';

const FillinInteractionNode = ({
  node,
  interactionState,
  onAttempt,
  shouldShowTransliteration,
  toggleTransliterationVisibility,
  learnLanguageAlpha3,
  translations,
  onRetry,
  onProceed
}) => {
  const gapRef = useRef();
  const { hintLevel, nextHintLevel, onHint } = useInputHint();

  const transliterationEnabled = features.get('is_transliteration');

  const [
    onTransliteration,
    activeTransliterations
  ] = useTransliterationUIState();

  const [handleKeyPress, isTransliterationLoading] = useTransliterationReplace(
    learnLanguageAlpha3,
    onTransliteration,
    transliterationEnabled
  );

  const handleHint = useCallback(() => {
    gapRef.current?.focus();
    onHint();
  }, [gapRef, onHint]);

  const handleRetry = useCallback(() => {
    gapRef.current?.deleteAll();
    onRetry();
  }, [gapRef, onRetry]);

  const handleToggleTransliterationVisibility = useCallback(() => {
    gapRef.current?.focus();
    toggleTransliterationVisibility();
  }, [gapRef, toggleTransliterationVisibility]);

  const isTransliterationTableSupported =
    transliterationEnabled && !isTransliterationLoading;

  return <>
    <Text color="spaceGray">
      <Gap
        ref={gapRef}
        submitHint={translations.submitHint}
        active={interactionState !== INTERACTION_STATES.INACTIVE}
        targetChoices={node.targetChoices}
        attempt={node?.attempt}
        handleKeyPress={handleKeyPress}
        interactionState={interactionState}
        hintLevel={hintLevel}
        learnLanguageAlpha3={learnLanguageAlpha3}
        onReturn={onAttempt}
        shouldShowTransliteration={shouldShowTransliteration}
        toggleTransliterationVisibility={handleToggleTransliterationVisibility}
        onRetry={onRetry}
      />
    </Text>

    {interactionState === INTERACTION_STATES.INPUT &&
      <>
        {isTransliterationTableSupported &&
        shouldShowTransliteration && (
          <TransliterationTable
            active={activeTransliterations}
            learnLanguageAlpha3={learnLanguageAlpha3}
          />
        )}

        {isiOS() ? (
          <NativeButtonSheet
            nextHintLevel={nextHintLevel}
            onHint={handleHint}
          />
        ) : (
          <WebButtonSheet
            nextHintLevel={nextHintLevel}
            onHint={handleHint}
            isTransliterationTableSupported={isTransliterationTableSupported}
            onTransliterationToggled={handleToggleTransliterationVisibility}
          />
        )}
      </>
    }

    {interactionState === INTERACTION_STATES.FEEDBACK &&
      <FillinFeedback
        attempt={node?.attempt}
        learnLanguageAlpha3={learnLanguageAlpha3}
        onRetry={handleRetry}
        onProceed={onProceed}
      />
    }
  </>;
};

FillinInteractionNode.propTypes = {
  node: interactiveNodeType.isRequired,
  attempt: attemptType,
  toggleTransliterationVisibility: PropTypes.func.isRequired,
  interactionState: PropTypes.string.isRequired,
  shouldShowTransliteration: PropTypes.bool.isRequired,
  onAttempt: PropTypes.func.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  onRetry: PropTypes.func.isRequired,
  onProceed: PropTypes.func.isRequired,
  translations: PropTypes.shape({
    submitHint: PropTypes.string.isRequired
  }).isRequired
};

const getTranslations = (translate) => {
  return {
    submitHint: translate('b3.pop_over.enter_reminder_content')
  };
};

export default withTranslations(getTranslations)(FillinInteractionNode);
