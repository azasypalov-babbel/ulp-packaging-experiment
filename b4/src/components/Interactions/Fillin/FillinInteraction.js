import React, { useCallback, useEffect, useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import useScene, { INTERACTION_STATES } from '../shared/scene/useScene';
import { matchResult } from '../shared/Write/helpers';
import FillinInteractionNode from './FillinInteractionNode';
import { toggleTransliterationVisibility } from '../../../dux/session/actions';
import Text from '../../shared/Text';
import { getActiveNode, getItem, getMistakeCount, isGap } from '../shared/scene/selectors';
import useHasLatestAttempt from '../shared/Write/hasLatestAttempt';
import { play, requestPrompt } from '../shared/scene/actions';
import useFeedbackSounds from '../shared/useFeedbackSounds';

const quantiseStates = ({ active, activeNode, feedbackDismissed }) => {
  if (!active || !activeNode) return INTERACTION_STATES.INACTIVE;

  const { number, pending } = activeNode.attempt;
  if (number && !pending && !feedbackDismissed) return INTERACTION_STATES.FEEDBACK;

  return INTERACTION_STATES.INPUT;
};

const FillinInteraction = ({
  item,
  active,
  onFinish,
  learnLanguageAlpha3,
  shouldShowTransliteration,
  toggleTransliterationVisibility,
  onAttempt
}) => {
  const [
    scene,
    dispatch
  ] = useScene({
    learnLanguageAlpha3,
    item,
    matchResult
  });

  const [playFeedback, isFeedbackSoundDone] = useFeedbackSounds({
    resetDependencies: [getActiveNode(scene)]
  });
  const [feedbackDismissed, setFeedbackDismissed] = useState(true);

  const { nodes } = getItem(scene);
  const activeNode = getActiveNode(scene);

  const interactionState = quantiseStates({ active, activeNode, feedbackDismissed });

  const latestAttempt = useHasLatestAttempt(activeNode && activeNode.attempt);

  const mistakes = getMistakeCount(scene);

  const handleAttempt = useCallback((attemptText) => {
    if (feedbackDismissed === true) {
      setFeedbackDismissed(false);
      dispatch(requestPrompt(attemptText));
    }
  }, [dispatch, feedbackDismissed]);

  const handleRetry = useCallback(() => {
    setFeedbackDismissed(true);
  }, []);

  const handleProceed = useCallback(() => {
    setFeedbackDismissed(true);
    dispatch(play());
  }, [dispatch]);

  useEffect(() => {
    if (latestAttempt && active) {
      onAttempt({ attempt: latestAttempt });
      playFeedback(latestAttempt.solved);
    }
  }, [latestAttempt, onAttempt, active, playFeedback]);

  useEffect(() => {
    if (!activeNode) {
      dispatch(play());
    }
  }, [activeNode, dispatch]);

  useEffect(() => {
    if (scene.completed && active && isFeedbackSoundDone) {
      onFinish(mistakes);
    }
  }, [scene.completed, onFinish, mistakes, isFeedbackSoundDone, active]);

  return (
    <div>
      {nodes.map((node, idx) => isGap(node)
        ? <FillinInteractionNode
          key={idx}
          node={node}
          learnLanguageAlpha3={learnLanguageAlpha3}
          interactionState={node.active ? interactionState : INTERACTION_STATES.INACTIVE}
          onAttempt={handleAttempt}
          onRetry={handleRetry}
          onProceed={handleProceed}
          toggleTransliterationVisibility={toggleTransliterationVisibility}
          shouldShowTransliteration={shouldShowTransliteration}
        />
        : <Text key={idx} color="spaceGray" dangerouslySetInnerHTML={{ __html: node.text }} />
      )}
    </div>
  );
};

const mapStateToProps = ({ session }) => ({
  learnLanguageAlpha3: session.learnLanguageAlpha3,
  shouldShowTransliteration: session.transliterationSettings.shouldShow
});

const mapDispatchToProps = {
  toggleTransliterationVisibility
};

FillinInteraction.propTypes = {
  item: PropTypes.object.isRequired,
  toggleTransliterationVisibility: PropTypes.func.isRequired,
  active: PropTypes.bool.isRequired,
  shouldShowTransliteration: PropTypes.bool.isRequired,
  onFinish: PropTypes.func.isRequired,
  onAttempt: PropTypes.func.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired
};

FillinInteraction.trainerConfig = {

};

export default connect(mapStateToProps, mapDispatchToProps)(FillinInteraction);
