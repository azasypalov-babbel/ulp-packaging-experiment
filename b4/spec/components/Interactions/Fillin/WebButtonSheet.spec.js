import React from 'react';
import { WebButtonSheet } from '../../../../src/components/Interactions/Fillin/WebButtonSheet';
import { shallow } from 'enzyme';

describe('<WebButtonSheet />', () => {
  const props = {
    onTransliterationToggled: () => {},
    track: () => {},
    isTransliterationTableSupported: true,
    nextHintLevel: 'HINT',
    onHint: () => {},
    translations: {
      typingTips: 'typing tips',
      solution: 'solution',
      hint: 'hint'
    }
  };

  const wrapper = shallow(<WebButtonSheet {...props}/>);

  test('it renders', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
