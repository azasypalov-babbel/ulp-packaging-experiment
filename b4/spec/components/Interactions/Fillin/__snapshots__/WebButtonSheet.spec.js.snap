// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WebButtonSheet /> it renders 1`] = `
<styled.div
  onClick={[Function]}
>
  <div>
    <Styled(WithKeypress(IconButton))
      data-selector="transliteration-toggle"
      iconName="KeyboardIcon"
      onClick={[Function]}
    >
      typing tips
    </Styled(WithKeypress(IconButton))>
  </div>
  <div>
    <Styled(WithKeypress(IconButton))
      data-selector="show-hint"
      iconName="LightBulbIcon"
      onClick={[Function]}
    >
      hint
    </Styled(WithKeypress(IconButton))>
  </div>
</styled.div>
`;
